sdk/lib/drivers/sound/mmixer/sup.c

```diff
@@ -689,21 +673,20 @@ MMixerSetGetVolumeControlDetails(
     if (!Input)
         return MM_STATUS_UNSUCCESSFUL; /* To prevent dereferencing NULL */
 
+    /* Get maximum available range */
+    MaxRange = VolumeData->SignedMaximum - VolumeData->SignedMinimum;
+
+    /* Get ratio of the hardware range (DB) to the logical units range */
+    Ratio = (float)MaxRange / 0x10000;
+
     /* Loop for each channel */
     for (Channel = 0; Channel < MixerControlDetails->cChannels; Channel++)
     {
         if (bSet)
         {
             /* FIXME SEH */
-            Index = Input[Channel].dwValue / VolumeData->InputSteppingDelta;
-
-            if (Index >= VolumeData->ValuesCount)
-            {
-                DPRINT1("Index %u out of bounds %u \n", Index, VolumeData->ValuesCount);
-                return MM_STATUS_INVALID_PARAMETER;
-            }
-
-            Value = VolumeData->Values[Index];
+            /* Convert from the logical units to the hardware range (DB) */
+            Value = Input[Channel].dwValue * Ratio - MaxRange + Ratio;
```

Comment:
Could you test:
```suggestion
            Value = (Input[Channel].dwValue + 1) * Ratio - MaxRange;
```
or:
```suggestion
            Value = (Input[Channel].dwValue + 1) * MaxRange / 0x10000 - MaxRange;
```
that would avoid usage of floats.

Reply:
@HBelusca just tested both your suggestions (the 2nd one from here and the one from below).
Unfortunately, they don't work properly at all. There are two cases when either input value (0 - 65535) is zero on set (means the most minimum), or an output one is zero on get (usually means the most maximum, if the audio miniport driver does not define a positive values in its volume DB range).
As you can test it with 0 (zero) for `Input[Channel].dwValue` and `Value`, if 0 is passed in at least one of the cases, it causes wrong calculation results and hence won't get us what we expect.
Also, using floats is badly required for 100% exactness, which is needed for the correct outcome result.

Reply:
For example, as it is for Intel AC97 driver, position is at the most minimum, input `dwValue` is 0. 0 + 1 = 1, then 1 * 3047424 (this is the max range for AC97) = 3047424, then 3047424 / 65536 (0x10000 in hexadecimal) = 46,5. Finally, 46,5 - 3047424 = -3047377,5. OK, it's probably almost what we expect, but to get exactly correct value, we need to subtract the resulting constant from the last resulting value. We can do it ONLY while we have a float variable, which stores the value, since that constant needs to be re-used anyway. :man_shrugging: 
It doesn't work correctly even for "get" case too. Let's imagine 0 is received after the call (which means the most maximum, so we need to get 65535 as result). Then, 0 + 3047424 = 3047424, then 3047424 * 65536 = 199715979264, then 199715979264 / 3047424 = 65536. Finally, 65536 - 1 = 65535. It works correctly for the maximum position of the slider, but it doesn't for the minimum, because minimum is -3047424 (again, defined by AC97), so the first and all further operations except last will result zero. That's what we need, but then we subtract 1, which will give us wrong value and break all further calculations again. We can avoid that by forcing to - 0 if the `Value` equals to `VolumeData->SignedMinimum`, but then it doesn't work anyway, because floats aren't used neither on set, nor on get.
**The important note here is, "get" case only retrieves the value previously set via "set" case. So, if the wrong one was set (e. g., -3047377,5, without the constant offset, the result is also wrong, and then, all further calculations are broken and so on, which leads to incorrect behaviour of the volume slider.** The slider position and the loudness of the sound directly depend on which value is set or get.

Reply:
UPDATE: just tested: even if I additionally subtract `MaxRange / 0x10000` in the formula for the "set" case, it doesn't work unlike I subtract the float variable. Even if I'll try to do it separately below, it won't work too, because again, the values are integer, so the values are not exact as they should be. By avoiding floats usage, we lost an exactness of the calculations, and hence it doesn't work correctly as we expect.